### PR TITLE
Added randomization to constant time swap in ECC to mitigate new Nonce@Once attack

### PR DESCRIPTION
**Pull request by milosprv**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/bn/bn_lib.c [hunk 1]
```diff
@@ -882,6 +882,84 @@ void BN_consttime_swap(BN_ULONG condition, BIGNUM *a, BIGNUM *b, int nwords)
...
+ * does a lot of XORs that have one operand ('t' in the code below) at zero,
+ * while condition==1 results in 't' having about 50-50 zeroes and ones.
+ * This bias makes the swap vulnerable to EM (and likely other) physical
+ * side-channel attacks like Nonce@Once
+ * (for details, see: "Nonce@Once: A Single-Trace EM Side Channel Attack on
+ *  Several Constant-Time Elliptic Curve Implementations in Mobile Platforms"
+ * by M. Alam, B. Yilmaz, F. Werner, N. Samwel, A. Zajic, D. Genkin, Y. Yarom,
+ * and M. Prvulovic, in IEEE Euro S&P 2021)
+ *
+ * r is the value to use for randomization during the swap
```
Comment (mattcaswell) [inline crypto/bn/bn_lib.c:899]: You talk about `r` here, but the function below seems to have `rand1` and `rand2`

File: crypto/bn/bn_lib.c [hunk 2]
```diff
@@ -882,6 +882,84 @@ void BN_consttime_swap(BN_ULONG condition, BIGNUM *a, BIGNUM *b, int nwords)
...
+ * nwords is the number of words to swap.
+ * Assumes that at least nwords are allocated in both a and b.
+ * Assumes that no more than nwords are used by either a or b.
+ */
+void BN_consttime_swap_randomized(BN_ULONG condition, BIGNUM *a, BIGNUM *b, int nwords, BN_ULONG rand1, BN_ULONG rand2)
```
Comment (mattcaswell) [inline crypto/bn/bn_lib.c:904]: Why do we need rand1 and rand2. AFAICT they are always set to the same thing by the caller.

File: crypto/ec/ec_mult.c [hunk 3]
```diff
@@ -236,6 +238,21 @@ int ec_scalar_mul_ladder(const EC_GROUP *group, EC_POINT *r,
         }
     }
 
+    /*
+     * Create random values for BN_consttime_swap_randomized, and for
+	 * constant-time swapping of Z_is_one in EC_POINT_SWAP below
+	 * This randomizatin is needed to prevent side channel attacks
```
Comment (mattcaswell) [inline crypto/ec/ec_mult.c:244]: Something odd happening with the indentation here

Comment (mattcaswell) [reply crypto/ec/ec_mult.c:244]: Typo: randomizatin

File: crypto/ec/ec_mult.c [hunk 4]
```diff
@@ -236,6 +238,21 @@ int ec_scalar_mul_ladder(const EC_GROUP *group, EC_POINT *r,
...
+     * (for details, see: "Nonce@Once: A Single-Trace EM Side Channel Attack on
+     *  Several Constant-Time Elliptic Curve Implementations in Mobile Platforms"
+     * by M. Alam, B. Yilmaz, F. Werner, N. Samwel, A. Zajic, D. Genkin, Y. Yarom,
+     * and M. Prvulovic, in IEEE Euro S&P 2021)
+     */
+    BN_ULONG *swap_rand=OPENSSL_malloc((cardinality_bits+2)*sizeof(BN_ULONG));
```
Comment (mattcaswell) [inline crypto/ec/ec_mult.c:250]: Don't mix declarations and code. Please declare `swap_rand` at the beginning of the function

Comment (mattcaswell) [reply crypto/ec/ec_mult.c:250]: Code style: spaces around operators( =, +,*), i.e.

````
    BN_ULONG *swap_rand = OPENSSL_malloc((cardinality_bits + 2) * sizeof(BN_ULONG));
````

File: crypto/ec/ec_mult.c [hunk 5]
```diff
@@ -236,6 +238,21 @@ int ec_scalar_mul_ladder(const EC_GROUP *group, EC_POINT *r,
...
+    if(RAND_bytes_ex(bn_get_libctx(ctx),(unsigned char *)swap_rand,(cardinality_bits+2)*sizeof(BN_ULONG))<=0) {
```
Comment (mattcaswell) [inline crypto/ec/ec_mult.c:251]: `bn_get_libctx` is called `ossl_bn_get_libctx` in the latest master so this is causing the CIs to fail.

Comment (mattcaswell) [reply crypto/ec/ec_mult.c:251]: Code style: spaces around operators

File: crypto/ec/ec_mult.c [hunk 6]
```diff
@@ -250,7 +267,7 @@ int ec_scalar_mul_ladder(const EC_GROUP *group, EC_POINT *r,
      * k := scalar + 2*cardinality
      */
     kbit = BN_is_bit_set(lambda, cardinality_bits);
-    BN_consttime_swap(kbit, k, lambda, group_top + 2);
+    BN_consttime_swap_randomized(kbit, k, lambda, group_top + 2, swap_rand[cardinality_bits],swap_rand[cardinality_bits]);
```
Comment (mattcaswell) [inline crypto/ec/ec_mult.c:270]: Code style: space after ,

File: crypto/ec/ec_mult.c (hunk 4 above)
Comment (mattcaswell) [reply crypto/ec/ec_mult.c:250]: Probably we need to check for a NULL return from OPENSSL_malloc

File: include/openssl/bn.h [hunk 7]
```diff
@@ -337,6 +337,7 @@ BIGNUM *BN_mod_sqrt(BIGNUM *ret,
                     const BIGNUM *a, const BIGNUM *n, BN_CTX *ctx);
 
 void BN_consttime_swap(BN_ULONG swap, BIGNUM *a, BIGNUM *b, int nwords);
+void BN_consttime_swap_randomized(BN_ULONG swap, BIGNUM *a, BIGNUM *b, int nwords, BN_ULONG rand1, BN_ULONG rand2);
```
Comment (mattcaswell) [inline include/openssl/bn.h:340]: This new public function needs to be documented. You also need to run "make update" to update the util/libcrypto.num file.

**Comment by beldmit** [comment]
Could you also please provide a link to the attack description?

File: crypto/bn/bn_lib.c (hunk 2 above)
Comment (mattcaswell) [reply crypto/bn/bn_lib.c:904]: Is it important that different callers can use a different rand value? Or could we generate a sufficiently large rand value once at startup and have all callers use the same one?

File: include/openssl/bn.h (hunk 7 above)
Comment (mattcaswell) [reply include/openssl/bn.h:340]: I'd like to see a test for this new function

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by kroeckx** [comment]
Is there a reason for it to be a new public function? Can't the exisitng function generate a random number itself?

**Comment by paulidale** [comment]
At the risk of being flamed, do we need cryptographically secure random bytes for this?

A fast non-cryptographic RNG would allow us to always blind the swapped values and close a whole suite of potential attacks with minimal performance loss.  Non-cryptographic blinding will be at least as good as none and it should still mitigate the attack here.

Paranoia would suggest not doing this of course.

File: crypto/bn/bn_lib.c (hunk 2 above)
Comment (richsalz) [reply crypto/bn/bn_lib.c:904]: Reply here just for context: seems rand1/rand2 are a holdover.  Removed rand2 and renamed rand1 to r.

File: crypto/ec/ec_mult.c (hunk 3 above)
Comment (richsalz) [reply crypto/ec/ec_mult.c:244]: fixed typo and indentation (tab/space) in the new PR.

File: crypto/ec/ec_mult.c (hunk 4 above)
Comment (richsalz) [reply crypto/ec/ec_mult.c:250]: fixing style, moving the decl up, and adding NULL check in the new PR.

File: include/openssl/bn.h (hunk 7 above)
Comment (richsalz) [reply include/openssl/bn.h:340]: For a new PR, what would a test of this function look like?

**Comment by milosprv** [comment]
For rand1 and rand2 - they need to be the same when calling the function,
but the function needs to use them as separate parameters  - if we use just
one parameter for both, the compiler optimizes the mitigation out. With
rand1 and rand2, when compiling the function, the compiler doesn't know
they are the same so it does not remove the mitigation.

Milos

**Comment by milosprv** [comment]
I agree, a pseudo-random number would do the trick - but I'm not really an
expert on OpenSSL's code, so what function would you suggest?

Milos

**Comment by milosprv** [comment]
They are separate args because the compiler can optimize out the mitigation
if rand1 and ran2 are the same variable. The optimize-out was happening at
least on the x86 default compilation. Because only intraprocedural
optimization is done on this function, by using rand1 and rand2 that are
actually the same when called, the compiler does not know that so it
retains the XORs that are needed for mitigation.

Milos

**Comment by paulidale** [comment]
> For rand1 and rand2 - they need to be the same when calling the function, but the function needs to use them as separate parameters - if we use just one parameter for both, the compiler optimizes the mitigation out. With rand1 and rand2, when compiling the function, the compiler doesn't know they are the same so it does not remove the mitigation.

This would be great added as a comment in the code.


> I agree, a pseudo-random number would do the trick - but I'm not really an expert on OpenSSL's code, so what function would you suggest? Milos

We've got a low quality RNG lying around in `impl_cache_flush_cache()` in `crypto/property/property.c` which would work but would require a bit of effort to extricate and have APIs added.  I'd suggest sticking with the current setup with a comment noting that low quality RNG would also work.

**Comment by richsalz** [comment]
So should this be er-oepned and my revision at #15702 be closed?  I can do the rand1/rand2/comment thing if desired.

**Comment by kaduk** [comment]
> > For rand1 and rand2 - they need to be the same when calling the function, but the function needs to use them as separate parameters - if we use just one parameter for both, the compiler optimizes the mitigation out. With rand1 and rand2, when compiling the function, the compiler doesn't know they are the same so it does not remove the mitigation.
> 
> This would be great added as a comment in the code.

I'm not sure to what extent we should rely on the current behavior of optimizing compilers to be retained through to the future; compiler authors tend to add more optimizations over time.

**Comment by richsalz** [comment]
Yes, compilers add more optimization, but if a current optimizer wipes away some code, then we should still protect against it. We've done that before, e.g, with zeroizing.

**Comment by richsalz** [comment]
I updated #15702 to have both parameters and explain why.

**Comment by richsalz** [comment]
@milosprv I started with this "fork of a PR" because I know that dealing with the project's source formatting nits can be a pain, and that I thought this was important to get into 3.0. Sadly, that seems unlikely.

**Comment by milosprv** [comment]
Please do add any changes you deem needed - there are a number of things
that I don't know how to handle. E.g. there is a request for a test suite
because there is a new function. But I didn't modify the existing
function because I don't know where else it is used (and if it is used,
whether it needs the randomization), and the new function really makes no
sense except in the context in which it is used (so does it really need a
separate test suite?). So please do make changes to the code so that it can
make it into a released version...

Thanks,
  Milos

**Comment by richsalz** [comment]
@milosprv if you look at the comments near the end of https://github.com/openssl/openssl/pull/15702 you can see that the team has some hesitancy about including this in 3.0  I am happy to do whatever is needed if they'll consider it for 3.0  I don't know if they will.  Perhaps you could comment over at that PR?

**Pull request closed without merging**